### PR TITLE
Feature/swagger joi

### PR DESCRIPTION
**Pull request by medhat93**
https://github.com/nodearch/rest/issues/2

### CODE REVIEW

File: src/swagger/open-api-schema.spec.ts [hunk 1]
```diff
@@ -0,0 +1,276 @@
+import { describe, it } from 'mocha';
```
Comment (AhmedAli7O1) [inline src/swagger/open-api-schema.spec.ts]: we need e2e testing for swagger it's very important

File: src/fileUpload/index.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+}
+
+function moveFilesToBody(req: Request) {
+  if (req.file && req.file.fieldname) {
+    req.body[req.file.fieldname] = req.file;
+    delete req.file;
+  }
+
+  if (req.files) {
```
[the remark this answers is not in the document]
Comment (medhat93) [reply src/fileUpload/index.ts]: no

File: src/route/route.info.ts [hunk 3]
```diff
@@ -29,6 +30,10 @@ export class RouteInfo {
       this.middlewares.push(validationMiddleware);
     }
 
+    if (fileUploadMiddleware) {
+      this.middlewares = [fileUploadMiddleware].concat(this.middlewares);
+    }
```
[the remark this answers is not in the document]
Comment (medhat93) [reply src/route/route.info.ts]: to make `fileUploadMiddleware` 1st middleware, as u maybe add some middlewares, guards or validations depends on files in req

File: src/swagger/open-api-schema.spec.ts (hunk 1 above)
Comment (medhat93) [reply src/swagger/open-api-schema.spec.ts]: check `server.e2e-spec` file, it have test cases for all rest features including swagger & file upload

File: src/server.ts [hunk 4]
```diff
@@ -50,10 +50,9 @@ export class RestServer implements IAppExtension {
 
     this.registerSequenceMiddlewares(this.expressSequence.slice(eRegisterRoutesIndex + 1, eStartIndex));
 
-    if (this.swagger && this.swagger.enable) {
+    if (this.swagger && this.swagger.path) {
       const swagger = new OpenApiSchema(controllers, this.swagger.options, this.joiValidationOptions);
-      await swagger.writeOpenAPI();
-      await swagger.register(this.expressApp);
+      await swagger.writeOpenAPI(this.swagger.path);
```
Comment (AhmedAli7O1) [inline src/server.ts]: we don't need this line either. we're going to trigger this from the CLI

**Pull request merged**